src/audio_core/hle/wmf_decoder_utils.cpp

```diff
@@ -0,0 +1,366 @@
+    IMFMediaType* t;
+
+    // If you know what you need and what you are doing, you can specify the condition instead of
+    // searching but it's better to use search since MFT may or may not support your output
+    // parameters
+    for (DWORD i = 0;; i++) {
+        hr = transform->GetOutputAvailableType(out_stream_id, i, &t);
+        if (hr == MF_E_NO_MORE_TYPES || hr == E_NOTIMPL) {
+            return 0;
+        }
+        if (FAILED(hr)) {
+            ReportError("failed to get output types for MFT", hr);
+            return -1;
+        }
+
+        hr = t->GetUINT32(MF_MT_AUDIO_BITS_PER_SAMPLE, &tmp);
+
+        if (FAILED(hr))
+            continue;
+        // select PCM-16 format
+        if (tmp == 32) {
+            hr = t->SetUINT32(MF_MT_AUDIO_BLOCK_ALIGNMENT, 1);
+            if (FAILED(hr)) {
+                ReportError("failed to set MF_MT_AUDIO_BLOCK_ALIGNMENT for MFT on output stream",
+                            hr);
+                return -1;
+            }
+            hr = transform->SetOutputType(out_stream_id, t, 0);
+            if (FAILED(hr)) {
+                ReportError("failed to select output types for MFT", hr);
+                return -1;
+            }
+            return 0;
+        } else {
+            continue;
+        }
+
+        return -1;
+    }
+
+    ReportError("MFT: Unable to find preferred output format", E_NOTIMPL);
+    return -1;
+}
+
+int detect_mediatype(char* buffer, size_t len, ADTSData* output, char** aac_tag) {
+    if (len < 7) {
+        return -1;
+    }
+
+    ADTSData tmp;
+    UINT8 aac_tmp[] = {0x01, 0x00, 0xfe, 00, 00, 00, 00, 00, 00, 00, 00, 00, 0x00, 0x00};
+    uint16_t tag = 0;
+
+    uint32_t result = parse_adts(buffer, &tmp);
+    if (result == 0) {
+        return -1;
+    }
+
+    tag = mf_get_aac_tag(tmp);
+    aac_tmp[12] |= (tag & 0xff00) >> 8;
+    aac_tmp[13] |= (tag & 0x00ff);
+    memcpy(*aac_tag, aac_tmp, 14);
+    memcpy(output, &tmp, sizeof(ADTSData));
+    return 0;
+}
+
+int mf_flush(IMFTransform** transform) {
+    HRESULT hr = (*transform)->ProcessMessage(MFT_MESSAGE_COMMAND_FLUSH, 0);
+    if (FAILED(hr)) {
+        ReportError("MFT: Flush command failed", hr);
+    }
+    hr = (*transform)->ProcessMessage(MFT_MESSAGE_NOTIFY_END_OF_STREAM, 0);
+    if (FAILED(hr)) {
+        ReportError("Failed to end streaming for MFT", hr);
+    }
+
+    return 0;
+}
+
+int send_sample(IMFTransform* transform, DWORD in_stream_id, IMFSample* in_sample) {
+    HRESULT hr = S_OK;
+
+    if (in_sample) {
+        hr = transform->ProcessInput(in_stream_id, in_sample, 0);
+        if (hr == MF_E_NOTACCEPTING) {
+            return 1; // try again
+        } else if (FAILED(hr)) {
+            ReportError("MFT: Failed to process input", hr);
+            return -1;
+        } // FAILED(hr)
+    } else {
+        hr = transform->ProcessMessage(MFT_MESSAGE_COMMAND_DRAIN, 0);
+        // ffmpeg: Some MFTs (AC3) will send a frame after each drain command (???), so
```

Comment:
why `ffmpeg:`?

Reply:
This function is taken from a proposed patch of `ffmpeg`. While this is not the case with AAC, I figured I should leave the comment here.

Reply:
I think I don't need to do anything for this one. Marking as resolved